From cf9b23831b05e21a97627ec700594d42af800ab6 Mon Sep 17 00:00:00 2001
From: "H.J. Lu" <hjl.tools@gmail.com>
Date: Wed, 16 May 2018 09:03:45 -0700
Subject: [PATCH 3/6] x86-64: Use IFUNC strncat inside libc.so

Unlike i386, we can call hidden IFUNC functions inside libc.so since
x86-64 PLT is always PIC.

Tested on x86-64.

	* sysdeps/x86_64/multiarch/strncat-c.c (STRNCAT_PRIMARY): Removed.
	Include <string/strncat.c>.
	* sysdeps/x86_64/multiarch/strncat.c (__strncat): New strong
	alias.
	(__GI___strncat): New hidden alias.

(cherry picked from commit 04958880e04264da97873b4d41d9bc34567afaef)
---
 sysdeps/x86_64/multiarch/strncat-c.c | 3 +--
 sysdeps/x86_64/multiarch/strncat.c   | 5 +++++
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/sysdeps/x86_64/multiarch/strncat-c.c b/sysdeps/x86_64/multiarch/strncat-c.c
index db8ad9917c..93a7fab7ea 100644
--- a/sysdeps/x86_64/multiarch/strncat-c.c
+++ b/sysdeps/x86_64/multiarch/strncat-c.c
@@ -1,3 +1,2 @@
 #define STRNCAT __strncat_sse2
-#define STRNCAT_PRIMARY
-#include "string/strncat.c"
+#include <string/strncat.c>
diff --git a/sysdeps/x86_64/multiarch/strncat.c b/sysdeps/x86_64/multiarch/strncat.c
index 1267eb4972..841c165565 100644
--- a/sysdeps/x86_64/multiarch/strncat.c
+++ b/sysdeps/x86_64/multiarch/strncat.c
@@ -27,4 +27,9 @@
 # include "ifunc-unaligned-ssse3.h"
 
 libc_ifunc_redirected (__redirect_strncat, strncat, IFUNC_SELECTOR ());
+strong_alias (strncat, __strncat);
+# ifdef SHARED
+__hidden_ver1 (strncat, __GI___strncat, __redirect_strncat)
+  __attribute__((visibility ("hidden")));
+# endif
 #endif
-- 
2.37.2

